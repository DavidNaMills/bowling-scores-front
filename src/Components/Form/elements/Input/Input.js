import React from 'react';
import classes from './Input.module.scss';
import { CreateSharp, ErrorSharp } from '@material-ui/icons'

const iconOk = [classes.icon, classes.icon__required];
const iconErr = [classes.icon, classes.icon__error];

const Input = ({ changed, id, type = 'text', label = null, placeholder = '', error = [], isValid = false, isRequired = false }) => {
    let styles = [classes.input, classes.input__default]

    if (isValid) {
        styles = styles.concat(classes.input__valid);
    }

    if (error.length > 0) {
        styles = styles.concat(classes.input__error);
    }

    return (
        <div className={classes.input__label}>
            {label && <label htmlFor={id}>{label}</label>}
            <div>
                <input
                    className={styles.join(' ')}
                    id={id}
                    type={type}
                    onChange={e => changed(e, id)}
                    placeholder={placeholder}
                />

                {
                    error.length > 0
                        ? <ErrorSharp className={iconErr.join(' ')} />
                        : isRequired && <CreateSharp className={iconOk.join(' ')} />
                }
            </div>
            {error.length > 0 &&
                <div className={classes.input__errorContainer}>
                    {
                        error.map((x, i) => <p key={i} className={classes.input__errorContainer_errorMsg}>{x}</p>, 0)
                    }
                </div>
            }
        </div>
    )
}

export default Input;