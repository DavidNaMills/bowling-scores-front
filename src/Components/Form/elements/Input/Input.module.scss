@import '../../../../styles/colors.scss';
@import '../../../../styles/fonts.scss';

.input{
    border: none;
    padding: 5px;
    width: $input_width;
    
    &__label{
        margin-top: $input_spacing;
    }

    &__default{
        background: rgb(230, 230, 230);
    }

    &__valid{
        -webkit-box-shadow: $valid_input_shadow;
        -moz-box-shadow: $valid_input_shadow;
        box-shadow: $valid_input_shadow;
        border-bottom: solid 1px $btn_default_bkg;
    };
    
    &__error{
        -webkit-box-shadow: $error_input_shadow;
        -moz-box-shadow: $error_input_shadow;
        box-shadow: $error_input_shadow;
        border-bottom: solid 1px $btn_danger_bkg;
    }

    &__errorContainer{
        width: $input_width;
        padding: 5px;

        -webkit-box-shadow: $error_input_shadow;
        -moz-box-shadow: $error_input_shadow;
        box-shadow: $error_input_shadow;

        &_errorMsg{
            font-size: $error_size_small;
            color: $btn_danger_bkg;
        }
    }
}

.icon{
    margin-left: 5px;
    display: flex;
    justify-content: center;
    align-content: center;

    &__required{
        color: $btn_default_bkg_active_hover;
    }

    &__error{
        color: $btn_danger_bkg_active;
    }
}