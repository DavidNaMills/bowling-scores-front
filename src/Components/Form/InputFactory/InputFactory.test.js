import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Input from '../elements/Input/Input';
import Inputfactory from './InputFactory';

const testFn = jest.fn();
const testData = {
    id: 'password',
    config: {
        elementtype: 'input',
        elementConfig: {
            type: 'text',
            placeholder: 'user_name'
        },
        value: '',
        validation: {
            minLength: {
                req: 10,
                errMsg: 'login_err1'
            },
            maxLength: {
                req: 20,
                errMsg: 'login_err2'
            }
        },
        touched: false,
        isValid: false,
        shouldValidate: true,
        hasErr: []
    }
}

describe('<InputFactory /> test suite', () => {
    describe('Snapshot tests', ()=>{
        test('<InputFactory/> snapshot test', ()=>{
            const component = renderer.create(<Inputfactory config={testData.config} id={testData.id} changed={testFn} />);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();
        })
    })

    it('should generate an <Input/> element', () => {
        const wrapper = mount(<Inputfactory config={testData.config} id={testData.id} changed={testFn} />)
        expect(wrapper.find(Input).length).toBe(1);
    })



    //TODO: handle errors
    //TODO: add more components when needed
    //returns error
})