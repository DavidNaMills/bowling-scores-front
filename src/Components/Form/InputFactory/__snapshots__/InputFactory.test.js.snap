// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputFactory /> test suite Snapshot tests <InputFactory/> snapshot test 1`] = `
<div
  className="input__label"
>
  <label
    htmlFor="password"
  >
    user_name
  </label>
  <div>
    <input
      className="input input__default"
      id="password"
      onChange={[Function]}
      placeholder="user_name"
      type="text"
      value=""
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root icon icon__required"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM21.41 6.34l-3.75-3.75-2.53 2.54 3.75 3.75 2.53-2.54z"
      />
    </svg>
  </div>
</div>
`;
